File: crates/photo-tauri/src/lib/burst.ts
// 连拍分组纯逻辑（T0 批次）：按显示顺序的 CaptureMeta 数组，将相邻两项拍摄时间
// （EXIF dateTaken）差 ≤ gapMs 的归为同一连拍组。纯前端实现，无后端改动。
// 组信息以「输入数组下标 → { groupId, size, pos }」映射返回，只登记 size≥2 的组；
// 调用方按需映射到 captures.items 下标（如过滤+排序后的显示序，见 filter store）。
import type { CaptureMeta } from './bindings'

/** 连拍组成员信息：groupId 组标识、size 组总张数、pos 组内序号（0 起，按显示序） */
export interface BurstEntry {
  groupId: string
  size: number
  pos: number
}

/** 组信息映射：key = 输入数组下标（显示序位置）；仅含 size≥2 的组 */
export type BurstGroupMap = Map<number, BurstEntry>

/**
 * EXIF 拍摄时间 → 毫秒时间戳（UTC）。
 * 兼容两种形态：EXIF「2026:06:28 10:15:30」与 mock/ISO「2026-08-01T10:15:30」；
 * 字段间分隔符不限定（年份 4 位 + 月/日/时/分/秒各 2 位）。
 * 用 Date.UTC 构造：连拍分组只关心时间差，时区在差值中相消，不受本地时区影响。
 * 解析失败/字段越界（如 13 月）返回 null。
 * （filter.ts 的 DATE_TAKEN_RE 只做 YYYY-MM-DD 日期串提取，语义不同，不复用。）
 */
export function parseExifDate(s: string | null): number | null {
  if (!s) return null
  const m = /^(\d{4})\D(\d{2})\D(\d{2})\D(\d{2})\D(\d{2})\D(\d{2})$/.exec(s.trim())
  if (!m) return null
  const y = Number(m[1])
  const mo = Number(m[2])
  const d = Number(m[3])
  const h = Number(m[4])
  const mi = Number(m[5])
  const se = Number(m[6])
  // Date.UTC 对越界字段会静默进位（如 13 月 → 次年 1 月），需先拦截，避免错误分组
  if (mo < 1 || mo > 12 || d < 1 || d > 31 || h > 23 || mi > 59 || se > 59) return null
  return Date.UTC(y, mo - 1, d, h, mi, se)
}

/**
 * 连拍分组：按传入顺序（调用方传显示顺序）遍历，相邻两项 dateTaken 时间差
 * ≤ gapMs 归入同组；dateTaken 为 null 或解析失败 = 独立单张（不登记，且切断组链）。
 * 返回 Map<下标, BurstEntry>，只包含 size≥2 的组（单张不成组、空数组返回空 Map）。
 */
export function computeBurstGroups(items: CaptureMeta[], gapMs = 2000): BurstGroupMap {
  // 第一遍：切分组边界（start/end 半开区间，按下标）
  const bounds: { start: number; end: number }[] = []
  let start = -1
  let prev: number | null = null
  for (let i = 0; i < items.length; i++) {
    const t = parseExifDate(items[i].dateTaken)
    if (t === null) {
      // null/解析失败：结算当前组并切断链（前后两段即使时间接近也不合并）
      if (start >= 0) bounds.push({ start, end: i })
      start = -1
      prev = null
      continue
    }
    if (start < 0) {
      start = i
    } else if (prev !== null && Math.abs(t - prev) <= gapMs) {
      // 与上一项时间差在阈值内 → 延续当前组
    } else {
      bounds.push({ start, end: i })
      start = i
    }
    prev = t
  }
  if (start >= 0) bounds.push({ start, end: items.length })

  // 第二遍：只登记 size≥2 的组（groupId 按组序编号，size/pos 为组结算后的终值）
  const map: BurstGroupMap = new Map()
  bounds.forEach((b, gi) => {
    const size = b.end - b.start
    if (size < 2) return
    const groupId = `burst-${gi}`
    for (let i = b.start; i < b.end; i++) {
      map.set(i, { groupId, size, pos: i - b.start })
    }
  })
  return map
}

File: crates/photo-ui/src/model/burst.rs
//! 连拍分组纯逻辑（对应 §4.4）。

use chrono::{NaiveDate, TimeZone, Utc};
use photo_domain::CaptureMeta;
use std::collections::HashMap;

/// 连拍组成员信息
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BurstEntry {
    pub group_id: String,
    pub size: usize,
    pub pos: usize,
}

/// 连拍组映射：key = 输入数组下标（显示序位置）；仅含 size >= 2 的组
pub type BurstGroupMap = HashMap<usize, BurstEntry>;

/// EXIF 拍摄时间 -> 毫秒时间戳（UTC）
///
/// 兼容两种形态：EXIF「2026:06:28 10:15:30」与 ISO「2026-08-01T10:15:30」
pub fn parse_exif_date(s: Option<&str>) -> Option<i64> {
    let s = s?.trim();
    if s.len() < 19 {
        return None;
    }
    let b = s.as_bytes();
    // 检查 4 位年与 5 个 2 位数
    if !b[0..4].iter().all(u8::is_ascii_digit)
        || b[4].is_ascii_digit()
        || !b[5..7].iter().all(u8::is_ascii_digit)
        || b[7].is_ascii_digit()
        || !b[8..10].iter().all(u8::is_ascii_digit)
        || b[10].is_ascii_digit()
        || !b[11..13].iter().all(u8::is_ascii_digit)
        || b[13].is_ascii_digit()
        || !b[14..16].iter().all(u8::is_ascii_digit)
        || b[16].is_ascii_digit()
        || !b[17..19].iter().all(u8::is_ascii_digit)
    {
        return None;
    }

    let y: i32 = s[0..4].parse().ok()?;
    let mo: u32 = s[5..7].parse().ok()?;
    let d: u32 = s[8..10].parse().ok()?;
    let h: u32 = s[11..13].parse().ok()?;
    let mi: u32 = s[14..16].parse().ok()?;
    let se: u32 = s[17..19].parse().ok()?;

    // 字段边界拦截（月 1-12，日 1-31，时 0-23，分 0-59，秒 0-59）
    if !(1..=12).contains(&mo) || !(1..=31).contains(&d) || h > 23 || mi > 59 || se > 59 {
        return None;
    }

    let date = NaiveDate::from_ymd_opt(y, mo, d)?;
    let dt = date.and_hms_opt(h, mi, se)?;
    Some(Utc.from_utc_datetime(&dt).timestamp_millis())
}

/// 连拍分组：按传入顺序（显示序）遍历，相邻两项拍摄时间差 <= gap_ms 归同组
pub fn compute_burst_groups(items: &[CaptureMeta], gap_ms: i64) -> BurstGroupMap {
    struct Range {
        start: usize,
        end: usize,
    }

    let mut bounds: Vec<Range> = Vec::new();
    let mut start: Option<usize> = None;
    let mut prev_t: Option<i64> = None;

    for (i, item) in items.iter().enumerate() {
        let t = parse_exif_date(item.date_taken.as_deref());
        match t {
            None => {
                // null/解析失败：结算当前组并切断链
                if let Some(s) = start {
                    bounds.push(Range { start: s, end: i });
                }
                start = None;
                prev_t = None;
            }
            Some(curr_t) => {
                match start {
                    None => {
                        start = Some(i);
                    }
                    Some(s) => {
                        if let Some(pt) = prev_t {
                            if (curr_t - pt).abs() <= gap_ms {
                                // 延续当前组
                            } else {
                                bounds.push(Range { start: s, end: i });
                                start = Some(i);
                            }
                        } else {
                            bounds.push(Range { start: s, end: i });
                            start = Some(i);
                        }
                    }
                }
                prev_t = Some(curr_t);
            }
        }
    }

    if let Some(s) = start {
        bounds.push(Range {
            start: s,
            end: items.len(),
        });
    }

    let mut map = HashMap::new();
    for (gi, b) in bounds.into_iter().enumerate() {
        let size = b.end - b.start;
        if size < 2 {
            continue;
        }
        let group_id = format!("burst-{gi}");
        for i in b.start..b.end {
            map.insert(
                i,
                BurstEntry {
                    group_id: group_id.clone(),
                    size,
                    pos: i - b.start,
                },
            );
        }
    }

    map
}
